### PR TITLE
New option for using PVLPrinter

### PR DESCRIPTION
**Pull request by petravandenbos-utwente**
This pull request contains:
- small changes to JavaPrinter
- new class PVLPrinter (I copied a lot from JavaPrinter)
- a new option in Configuration for using PVLPrinter in Main (option usage: --session path-to-file)
- small bug fix in Type

### CODE REVIEW

File: col/src/main/java/vct/col/ast/print/JavaPrinter.java [hunk 1]
```diff
@@ -713,14 +714,19 @@ public void visit(DeclarationStatement s){
...
     }
     if (!in_expr) out.printf(";");
   }
 
   public void visit(Method m){
+    if (m.kind == Method.Kind.Constructor && m.getBody() instanceof BlockStatement) { //empty constructors don't need to be printed
+      if (((BlockStatement) m.getBody()).isEmpty()) {
+        return;
+      }
+    }
```
Comment (niomaster) [inline col/src/main/java/vct/col/ast/print/JavaPrinter.java:729]: See <https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.8.9>, we should also check:
* the number of arguments is 0
* the method throws nothing
* there is no other constructor

File: col/src/main/java/vct/col/ast/print/PVLPrinter.java [hunk 2]
```diff
@@ -0,0 +1,1532 @@
...
+        }
+        out.printf(")");
+        if(i==N-1){
+          out.lnprintf("];");
+        } else {
+          out.lnprintf(",");
+        }
+      }
+      out.lnprintf("  @ * /");
+    } else {*/
```
Comment (niomaster) [inline col/src/main/java/vct/col/ast/print/PVLPrinter.java:825]: let's remove this

File: col/src/main/java/vct/col/ast/print/PVLPrinter.java [hunk 3]
```diff
@@ -0,0 +1,1532 @@
...
+                nextExpr();
+                s.getGuard(i).accept(this);
+                out.lnprintf(") ");
+            }
+            s.getStatement(i).accept(this);
+            if (!self_terminating(s.getStatement(i))){
+                out.lnprintf(";");
+            }
+        }
+        //}
```
Comment (niomaster) [inline col/src/main/java/vct/col/ast/print/PVLPrinter.java:852]: idem

File: col/src/main/java/vct/col/ast/type/Type.scala [hunk 4]
```diff
@@ -24,7 +24,7 @@ abstract class Type(val args:List[ASTNode]) extends ASTNode {
   def argsJava = args.asJava
   
   /** Tests whether `args` has any arguments. */
-  def hasArguments = args.isEmpty
+  def hasArguments = !args.isEmpty
```
Comment (niomaster) [inline col/src/main/java/vct/col/ast/type/Type.scala:27]: Please use `args.nonEmpty`

File: src/main/java/vct/main/Main.java [hunk 5]
```diff
@@ -329,6 +329,10 @@ public static void main(String[] args) throws Throwable
       } else if (silver.used()||chalice.get()) {
         passes=new LinkedBlockingDeque<String>();
 
+        if(Configuration.session_file != null) {
```
Comment (niomaster) [inline src/main/java/vct/main/Main.java:332]: I think this should be `Configuration.session_file.get()`?

**Review by niomaster — CHANGES_REQUESTED** [review]

File: col/src/main/java/vct/col/ast/print/JavaPrinter.java (hunk 1 above)
Comment (petravandenbos-utwente) [reply col/src/main/java/vct/col/ast/print/JavaPrinter.java:729]: In method ```public void visit(Method m)``` I do not have information about other constructors. Do you want me to remove this if-statement, and just change back to the situation before this PR?

Comment (niomaster) [reply col/src/main/java/vct/col/ast/print/JavaPrinter.java:729]: I'm happy to assist in getting this to work better, but you can decide for yourself if it's worth it. With #541 we can probably also ensure that your formatting can happen before the default constructor is added instead.

Comment (petravandenbos-utwente) [reply col/src/main/java/vct/col/ast/print/JavaPrinter.java]: For now, I change it back to what it was before.

**Review by niomaster — APPROVED** [review]

**Pull request merged**